File: dart/packages/aoc2022/lib/src/day08.dart
// https://adventofcode.com/2022/day/8

import 'package:aoc/aoc.dart';
import 'package:aoc/util/collection.dart';
import 'package:aoc/util/grid.dart';
import 'package:aoc/util/string.dart';
import 'package:aoc/util/vec.dart';
import 'package:collection/collection.dart';

main() => Day08().solve();

class Day08 extends AdventDay {
  Day08() : super(2022, 8);

  @override
  dynamic part1(String input) {
    final grid = parseGrid(input);

    final Set<Vec> visible = {};
    for (int col = 0; col < grid.width; col++) {
      visible.addAll(seenInDirection(grid, Vec(col, 0), Vec.down));
      visible.addAll(seenInDirection(grid, Vec(col, grid.height - 1), Vec.up));
    }
    for (int row = 0; row < grid.height; row++) {
      visible.addAll(seenInDirection(grid, Vec(0, row), Vec.right));
      visible.addAll(seenInDirection(grid, Vec(grid.width - 1, row), Vec.left));
    }
    return visible.length;
  }

  @override
  dynamic part2(String input) {
    final grid = parseGrid(input);
    int score(Vec tree) =>
        Vec.orthogonalDirs.map((d) => numSeenFromTree(grid, tree, d)).product;
    return grid.locations().map(score).max;
  }

  Iterable<Vec> seenInDirection(Grid<int> grid, Vec starting, Vec dir) {
    final List<Vec> seen = [starting];
    Vec max = starting;
    Vec current = starting + dir;
    while (grid.validLocation(current)) {
      if (grid.value(max) < grid.value(current)) {
        seen.add(current);
        max = current;
      }
      current = current + dir;
    }
    return seen;
  }

  int numSeenFromTree(Grid<int> grid, Vec tree, Vec dir) {
    int seen = 0;
    int treeHeight = grid.value(tree);
    Vec current = tree + dir;
    while (grid.validLocation(current)) {
      seen += 1;
      if (grid.value(current) >= treeHeight) {
        return seen;
      }
      current = current + dir;
    }
    return seen;
  }

  Grid<int> parseGrid(String input) {
    final data =
        input.lines.map((l) => l.chars.map(int.parse).toList()).toList();
    return Grid<int>.from(data, 0);
  }
}

File: clojure/src/aoc2022/day08.clj
;; https://adventofcode.com/2022/day/8
(ns aoc2022.day08
  (:require
   [aoc.day :as d]
   [aoc.util.char :as char]
   [aoc.util.collection :as c]
   [aoc.util.grid :as g]
   [aoc.util.math :refer [product]]
   [aoc.util.pos :as p]))

(defn input [] (d/day-input 2022 8))

(defn- heights-from [grid pos dir]
  (->> (iterate (partial p/pos+ dir) pos)
       (map grid)
       (take-while identity)
       rest))

(defn- visible-in? [grid pos dir]
  (let [height (grid pos)]
    (every? #(> height %) (heights-from grid pos dir))))

(defn- visible? [grid pos]
  (some (partial visible-in? grid pos) p/orthogonal-dirs))

(defn- visible-trees [grid]
  (filter #(visible? grid %) (keys grid)))

(defn- scenic-score-from [grid pos dir]
  (let [height (grid pos)]
    (count (c/take-upto #(< % height) (heights-from grid pos dir)))))

(defn- scenic-score [grid pos]
  (->> p/orthogonal-dirs
       (map (partial scenic-score-from grid pos))
       product))

(defn part1 [input]
  (->> (g/->grid input char/digit)
       visible-trees
       count))

(defn part2 [input]
  (let [grid (g/->grid input char/digit)]
    (transduce (map (partial scenic-score grid)) max 0 (keys grid))))
